File: java/pattern/src/decorator/ICar.java
package decorator;

/**
 * 抽象构件
 */
public interface ICar {
    void move();
}

/**
 * 具体构件对象（真实对象）
 */
class Car implements ICar {

    @Override
    public void move() {
        System.out.println("陆地上跑");
    }
}

/**
 * Decorator 装饰角色
 */
class SuperCar implements ICar {
    protected ICar car;

    public SuperCar(ICar car) {
        this.car = car;
    }

    @Override
    public void move() {
        car.move();
    }
}


/**
 * ConcreteDecorator具体装饰角色
 */
class FlyCar extends SuperCar {

    public FlyCar(ICar car) {
        super(car);
    }

    public void fly() {
        System.out.println("天上飞！");
    }

    @Override
    public void move() {
        super.move();
        fly();
    }
}



/**
 * ConcreteDecorator具体装饰角色
 */
class WaterCar extends SuperCar {

    public WaterCar(ICar car) {
        super(car);
    }

    public void swim() {
        System.out.println("水上漂！");
    }

    @Override
    public void move() {
        super.move();
        swim();
    }
}


/**
 * ConcreteDecorator具体装饰角色
 */
class AICar extends SuperCar {

    public AICar(ICar car) {
        super(car);
    }

    public void autoMove() {
        System.out.println("智能驾驶，自动在陆地上跑！");
    }

    @Override
    public void move() {
        super.move();
        autoMove();
    }
}

File: php/decorator/ICar.php
<?php

/**
 * 抽象构件
 */
interface ICar
{
    public function move();
}

/**
 * 具体构件对象（真实对象）
 */
class Car implements ICar
{
    public function move()
    {
        echo '陆地上跑', PHP_EOL;
    }
}

/**
 * Decorator 装饰角色
 */
class SuperCar implements ICar
{
    protected $iCar;

    public function __construct($car)
    {
        $this->iCar = $car;
    }

    public function move()
    {
        $this->iCar->move();
    }
}

/**
 * ConcreteDecorator具体装饰角色
 */
class FlyCar extends SuperCar
{
    public function fly()
    {
        echo '天上飞', PHP_EOL;
    }

    public function move()
    {
        $this->iCar->move();

        $this->fly();
    }
}



/**
 * ConcreteDecorator具体装饰角色
 */
class WaterCar extends SuperCar
{
    public function water()
    {
        echo '水上漂', PHP_EOL;
    }

    public function move()
    {
        $this->iCar->move();

        $this->water();
    }
}



/**
 * ConcreteDecorator具体装饰角色
 */
class AICar extends SuperCar
{
    public function autoMove()
    {
        echo '智能驾驶，自动在陆地上跑！', PHP_EOL;
    }

    public function move()
    {
        $this->iCar->move();

        $this->autoMove();
    }
}
